bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+
+sealed class VarState(val onStack: Boolean, val locked: Boolean) {
+}
+object Declared      extends VarState(false, false)
+object Captured      extends VarState(false, false)
+object BoundFirst    extends VarState(true, true)
+object ToInstantiate extends VarState(false, true)
+
+case class VarAndState(v: Variable, state: VarState)
+
+/**
+ * Looks for declarations and the usages of those declarations.
+ * The Transform methods detect local statements and add declared variables to the environment
+ * The Walk methods are collecting their usages and tagging them
+ * 
+ * If a variable is bound after its declaration, we can shortcut it
+ * 
+ * When a proc is declared, it is difficult to track where it will end up. So if a call occurs,
+ * by precaution, all captured variable that were not bound should be instantiated
+ * 	This means that any A = {IntToFloat 3} will instantiate because it results in {IntToFloat 3 A}
+ *  that could potentially call end up calling A
+ *  
+ * When having two branches, if a variable does not end in the same state, it it better to instantiate it
+ * 
+ * If a variable is encountered, it is always put into the modified set (if any) except if it was already
+ * in a BoundFirst or a ToInstantiate state.
+ * 
+ * In try-catch or andthen/orelse nodes, variables that have not been bound yet end up in ToInstantiate state
+ *  
+ * The basic idea is that if no usage appears before binding, no need to create a variable.
+ */
+class OnStackMarking(val declVars: mutable.HashMap[Symbol, VarAndState],
+    val modified: Option[mutable.Set[Variable]]) extends Transformer with Walker {
+  def this(declVars: mutable.HashMap[Symbol, VarAndState]) = this(declVars, None)
+  def this() = this(mutable.HashMap())
+  
+  def tagVariable(v: Variable, state: VarState) = {
+    declVars.put(v.symbol, VarAndState(v, state))
+    modified.foreach { set => set.add(v) }
+  }
+  
+  /**
+   * Looks for affected variables using a sandboxed marker
+   */
+  def affectedVariables(affectedSet: mutable.Set[Variable], marking: (OnStackMarking)=>Unit): mutable.Set[Variable] = {
+    val affected = if (affectedSet == null) mutable.Set[Variable]() else affectedSet
+    val marker = new OnStackMarking(declVars.clone(), Some(affected)) {
+      // do nothing when encountering a "bond on stack" node. Bound and Used variables will get the same treatment
+      override def onStackBind(bind: BindCommon, v: Variable) = {}
+    }
+    marking(marker)
+    affected
+  }
+  
+  /**
+   * Runs the algorithm in sandboxed branches and merges the result in current marker
+   */
+  def conjugate[A](collector: mutable.HashMap[Symbol, VarAndState], branches: ((OnStackMarking)=>Unit)*): Unit = {
+    val collectorClone = collector.clone()
+    val modified = mutable.Set[Variable]()
+    val bindNodes = mutable.HashMap[Symbol, Seq[BindCommon]]()
+    for ((branch, i) <- branches.zipWithIndex) {
+      val current = if (i == 0) collector else collectorClone.clone() 
```

Comment:
Yes, the collector is the one on which the merging of the branching will be accessible. It will be incrementally accumulating information of all branches, but the information of the first branch can be fed as-is. It seemed easier for writing next lines.
The important thing is the collector is cloned before being affected by the first branch, so next branches see the state before the 1st one.